File: 3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    // hashmap & two pointers & queue: O(N)time & space, 25.91%, ok
    int lengthOfLongestSubstring(string s) {
        if(s.empty()) return 0;
        int ret = 0, start = 0;
        unordered_map<int, int> map;
        queue<char> record;
        for(int i = 0; i < s.size(); i++){
            record.push(s[i]);
            if(map.count(s[i])){
                ret = max(ret, i - start);
                start = map[s[i]] + 1;
                while(record.front() != s[start]){
                    map.erase(record.front());
                    record.pop();
                }
            }
            map[s[i]] = i;
        }
        ret = max(ret, int(s.size()) - start);
        return ret;
    }

    // hashmap & two pointers: O(N)time & space, 51.17%, ok
    int lengthOfLongestSubstring(string s) {
        if(s.empty()) return 0;
        int ret = 0, start = 0;
        unordered_map<int, int> map;
        for(int i = 0; i < s.size(); i++){
            if(map.count(s[i]) && map[s[i]] >= start){  // replace queue
                ret = max(ret, i - start);
                start = map[s[i]] + 1;
            }
            map[s[i]] = i;
        }
        ret = max(ret, int(s.size()) - start);
        return ret;
    }
};

File: 3. Longest Substring Without Repeating Characters.py
from collections import Counter
class Solution(object):
    
    # # brute force: O(N3) time, O(N) space, TLE
    # def lengthOfLongestSubstring(self, s):
    #     """
    #     :type s: str
    #     :rtype: int
    #     """
    #     ret = 0
    #     for i in xrange(len(s)):
    #         for j in xrange(i, len(s)):
    #             cnt = Counter(s[i:j + 1])
    #             if len(filter(lambda x: x > 1, cnt.values())) == 0:
    #                 ret = max(ret, j - i + 1)
    #     return ret
    
    # hash table: store the latest index of each char
    # O(N) time, O(N) space
    def lengthOfLongestSubstring(self, s):
        ret, i = 0, 0
        pos = {}
        for j in xrange(len(s)):
            if s[j] in pos and pos[s[j]] >= i:   # bug: pos[s[j]] >= i
                ret = max(ret, j - i)
                i = pos[s[j]] + 1
            pos[s[j]] = j   # save lastest position
        ret = max(ret, len(s) - i)
        print '----'
        return ret
        